### DISCUSSION TITLE
Is it possible to let vcpkg `git clone` instead of `download archive`?

### DISCUSSION THREAD

**Discussion by loynoir** [microsoft/vcpkg#37586] [General]
## Background

- I setup self hosted git mirror `self-hosted-git` for docs and cache performance.

- I update git config to redirect git clone `https://github.com/foo/bar` to `https://self-hosted-git/mirror/github-foo-bar`.

## Actual

Currently, when I install a package, `vcpkg` download archive from github, not using `git clone`.

```txt
Building bar[xxx,yyy]:x64-linux...
-- Downloading https://github.com/foo/bar/archive/1.2.3.tar.gz -> foo-bar-1.2.3.tar.gz...
-- Extracting source /var/cache/vcpkg/foo-bar-1.2.3.tar.gz
```

## Question

Is it possible to let vcpkg

- git clone `https://github.com/foo/bar`, or said `https://self-hosted-git/mirror/github-foo-bar`

instead of 

- `Downloading https://github.com/foo/bar/archive/1.2.3.tar.gz`?

Thanks.

**Comment by dg0yt** [comment]
`git clone` is incompatible with the regular vcpkg asset caching because asset caching relies on exact SHA512 of the content blobs.
And many ports must use regular tarballs because git is not available or not including some generated files.

(Did you consider configuring asset caching instead?)

**Reply by loynoir** [reply]
Thanks for the advice.

Using it now. :)

Still have interest in re-using git cache.

Because

- git mirror server have **builtin** support to update to new version in **idle time**, **ahead of the time** of vcpkg **busy time**.

- I don't know if there are exists tools for asset caching.

**Reply by dg0yt** [reply]
You can share and fill an asset cache across multiple installations and even computers. You can run `vcpkg install --only-downloads` ahead of time. etc.

> I don't know if there are exists tools for asset caching.

No idea what this means. It is just a storage adressed with `<SHA512>`. 
The reference may be a litte short:
https://learn.microsoft.com/en-us/vcpkg/users/assetcaching
But there is also
https://learn.microsoft.com/en-us/vcpkg/examples/asset-caching-source-nuget

(I don't used it. Locally, I have a growing `downloads` folder.)

**Reply by loynoir** [reply]
> > I don't know if there are exists tools for asset caching.
> 
> No idea what this means.

I mean

- git server auto update source code to new version

- seems like I may need to write some code to watch all `vcpkg.json`, and run them periodly, to update asset cache to new version

**Reply by dg0yt** [reply]
Hm, I really don't see what you gain over the existing asset caching and running `--only-downloads`. (Apart from full git mirroring for using only part of the history.)
And for the `vcpkg_from_github` ports, you can install with `--head` to have them `git clone` the latest git HEAD (not release) instead of downloading the tarball. But then you will stumble over the patches which no longer apply.
